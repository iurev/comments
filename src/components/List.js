import React, { Component } from 'react';
import styled from 'styled-components'
import { connect } from 'react-redux'
import { createSelector } from 'reselect'
import humanizeDuration from 'humanize-duration'
import Item from './Item'
import {
  feedbacksLimitedList,
} from '../selectors'
import actions from '../actions'

const Root = styled.section`
  display: flex;
  flex-direction: column;
  width: 100%;
  padding: 10px;
`

const selector = createSelector(
  feedbacksLimitedList,
  list => ({
    feedbacks: list.map((item) => {
      const { author, body, title, attachments, statistics, closed, last_activity_at, created_at } = item
      let avatar
      try {
        avatar = attachments[0].variants.profile.url
      } catch {}
      // TODO: add toCamelCase function to data
      const { vote_count, answer_count } = statistics
  
      const authorTime = humanizeDuration(new Date() - new Date(last_activity_at), { language: 'de', largest: 1 })
      const commentTime = humanizeDuration(new Date() - new Date(created_at), { language: 'de', largest: 1 })

      return {
        author: {
          ...author,
          avatar,
          time: authorTime,
        },
        comment: {
          title,
          text: body,
        },
        bottom: {
          closed,
          voteCount: vote_count,
          answerCount: answer_count,
          time: commentTime,
        },
      }
    })
  })
)

class List extends Component {
  componentDidMount() {
    this.props.load()
  }
  
  render() {
    return (
      <Root>
        {this.props.feedbacks.map((data, i) => <Item
          key={i}
          {...data}
        />)}
      </Root>
    );
  }
}

export default connect(
  state => selector(state),
  actions,
)(List);
