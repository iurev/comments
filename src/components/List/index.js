import React, { Component } from 'react';
import { connect } from 'react-redux'
import { createSelector } from 'reselect'
import humanizeDuration from 'humanize-duration'
import ListView from './view'
import {
  feedbacksLimitedList,
} from '../../selectors'
import actions from '../../actions'

const HUMANIZE_CONFIG = { language: 'de', largest: 1 }

const humanize = (date) => (
  humanizeDuration(new Date() - new Date(date), HUMANIZE_CONFIG)
)

const getAuthor = (item) => {
  const { author, lastActivityAt } = item
  let avatar
  try {
    avatar = author.avatar.variants.profile.url
  } catch {}
  return {
    ...author,
    avatar,
    time: humanize(lastActivityAt),
  }
}

const getComment = (item) => {
  const { rating, body, title, attachments } = item
  let image
  try {
    image = attachments[0].variants.profile.url
  } catch {}

  return {
    title,
    text: body,
    image,
    rating,
  }
}

const getBottom = (item) => {
  const { statistics, closed, createdAt } = item
  const { voteCount, answerCount } = statistics

  return {
    closed,
    voteCount,
    answerCount,
    time: humanize(createdAt),
  }
}

const selector = createSelector(
  feedbacksLimitedList,
  list => ({
    feedbacks: list.map((item) => {
      return {
        author: getAuthor(item),
        comment: getComment(item),
        bottom: getBottom(item),
      }
    })
  })
)

export class List extends Component {
  componentDidMount() {
    this.props.load()
  }

  render() {
    return (
      <ListView feedbacks={this.props.feedbacks} />
    );
  }
}

export default connect(
  state => selector(state),
  actions,
)(List);
