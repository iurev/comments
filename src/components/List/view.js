import React from 'react'
import styled from 'styled-components'
import Item from '../Item'


const Root = styled.section`
  display: flex;
  flex-direction: column;
  width: 100%;
  padding: 10px;
`

const ListView = ({ feedbacks }) => (
  <Root>
    {feedbacks.map(({ author, bottom, comment }, i) => (<Item
      key={i}
      author={author}
      bottom={bottom}
      comment={comment}
    />))}
  </Root>
)

export default ListView
