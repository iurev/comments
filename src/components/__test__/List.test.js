import React from 'react';
import ShallowRenderer from 'react-test-renderer/shallow';
import { List } from '../List';

it('empty feedbacks', () => {
  const tree = new ShallowRenderer().render(<List
    feedbacks={[]}
  />)
  expect(tree).toMatchSnapshot();
});

it('1 feedback', () => {
  const tree = new ShallowRenderer().render(<List
    feedbacks={[{}]}
  />)
  expect(tree).toMatchSnapshot();
});
