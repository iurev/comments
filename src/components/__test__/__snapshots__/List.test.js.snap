// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`1 feedback 1`] = `
<ForwardRef>
  <Item />
</ForwardRef>
`;

exports[`empty feedbacks 1`] = `<ForwardRef />`;
